tmail-backend/rate-limiter/rate-limiter-api/src/test/scala/com/linagora/tmail/rate/limiter/api/RateLimitationPlanRepositoryContract.scala

```diff
@@ -0,0 +1,180 @@
+  }
+
+  @Test
+  def updateShouldNotModifyAnotherEntry(): Unit = {
+    val rateLimitingPlan: RateLimitingPlan = SMono.fromPublisher(testee.create(CREATION_REQUEST))
+      .block()
+    val rateLimitingPlan2: RateLimitingPlan = SMono.fromPublisher(testee.create(CREATION_REQUEST))
+      .block()
+
+    SMono.fromPublisher(testee.update(RESET_REQUEST.copy(id = rateLimitingPlan.id))).block()
+
+    assertThat(SMono.fromPublisher(testee.get(rateLimitingPlan2.id))
+      .block())
+      .isEqualTo(rateLimitingPlan2)
+  }
+
+  @Test
+  def getShouldThrowWhenIdNotFound(): Unit = {
+    assertThatThrownBy(() => SMono.fromPublisher(testee.get(RateLimitingPlanId.generate)).block())
+      .isInstanceOf(classOf[RateLimitingPlanNotFoundException])
+  }
+
+  @Test
+  def getShouldReturnEntryWhenIdExists(): Unit = {
+    val rateLimitingPlan: RateLimitingPlan = SMono.fromPublisher(testee.create(CREATION_REQUEST))
+      .block()
+
+    assertThat(SMono.fromPublisher(testee.get(rateLimitingPlan.id)).block())
+      .isEqualTo(rateLimitingPlan)
+  }
+
+  @Test
+  def planExistsShouldReturnFalseByDefault(): Unit = {
+    assertThat(SMono.fromPublisher(testee.planExists(RateLimitingPlanId.generate)).block())
+      .isEqualTo(false)
+  }
+
+  @Test
+  def planExistsShouldReturnTrueWhenPlanExists(): Unit = {
+    val rateLimitingPlan: RateLimitingPlan = SMono.fromPublisher(testee.create(CREATION_REQUEST))
+      .block()
+
+    assertThat(SMono.fromPublisher(testee.planExists(rateLimitingPlan.id)).block())
+      .isEqualTo(true)
+  }
+
+  @Test
+  def listShouldReturnEmptyByDefault(): Unit = {
+    assertThat(SFlux.fromPublisher(testee.list()).collectSeq().block().asJava)
+      .isEmpty()
+  }
+
+  @Test
+  def listShouldReturnStoredEntriesWhenHasSingleElement(): Unit = {
+    val rateLimitingPlan: RateLimitingPlan = SMono.fromPublisher(testee.create(CREATION_REQUEST))
+      .block()
+
+    assertThat(SFlux.fromPublisher(testee.list()).collectSeq().block().asJava)
+      .containsExactlyInAnyOrder(rateLimitingPlan)
+  }
+
+  @Test
+  def listShouldReturnStoredEntriesWhenHasSeveralElement(): Unit = {
+    val rateLimitingPlan: RateLimitingPlan = SMono.fromPublisher(testee.create(CREATION_REQUEST))
+      .block()
+    val rateLimitingPlan2: RateLimitingPlan = SMono.fromPublisher(testee.create(CREATION_REQUEST.copy(name = RateLimitingPlanName("name2"))))
+      .block()
+
+    assertThat(SFlux.fromPublisher(testee.list()).collectSeq().block().asJava)
+      .containsExactlyInAnyOrder(rateLimitingPlan, rateLimitingPlan2)
+  }
+}
+
+class InMemoryRateLimitationPlanRepositoryTest extends RateLimitationPlanRepositoryContract {
+  var inMemoryRateLimitationPlanRepository: InMemoryRateLimitationPlanRepository = _
+
+  override def testee: RateLimitationPlanRepository = inMemoryRateLimitationPlanRepository
+
+  @BeforeEach
+  def beforeEach(): Unit = {
+    inMemoryRateLimitationPlanRepository = new InMemoryRateLimitationPlanRepository();
+  }
```

Comment:
```suggestion
  val inMemoryRateLimitationPlanRepository: InMemoryRateLimitationPlanRepository = new InMemoryRateLimitationPlanRepository()

  override def testee: RateLimitationPlanRepository = inMemoryRateLimitationPlanRepository
```

Reply:
IMO No
re-create an instance for each test will help more isolation

Reply:
I just debug and learn something.
With the code I suggested:
- val inMemoryRateLimitationPlanRepository means this variable belongs to the instance of InMemoryTest class and can not be changed in that instance life cycle. => still assure singleton in a single test.
- It looks like Junit Test create a new InMemoryTest instance for each test (without BeforeEach), so a new val inMemoryRateLimitationPlanRepository has been created again => still assure isolation.
 ![image](https://user-images.githubusercontent.com/55171818/154051766-91318fa1-4fa4-4349-9003-4a1e11ba7b2f.png)